app/models/concerns/variant_stock.rb

```diff
@@ -105,6 +105,7 @@ def on_demand=(new_value)
     # unknown reasons, it does not pass the test.
     stock_items.each do |item|
       item.backorderable = new_value
+      item.save
```

Comment:
I would really think twice to do this. These are the kind of things that cause the myriad of UPDATE and INSERT statements we see in development, resulting in multiple and pointless writes sometimes. 

This is just like another db column baked by AR and in that case, AR won't persist on assignments. I would not do it in our case either. To me, this is the responsibility of the consumer of this setter method.

Reply:
ok, thinking twice... I write up my thoughts so you can comment:

1. first I wanted to see **if this works**. answer: yes! this fixes not just the 21 one broken tests but a couple more.

2. then, I wanted to **understand** why: I think I now understand what's going on, if we dont save on_demand, the call to on_hand is saving the stock_item and ignoring the on_demand value (stock_item.backorderable) set just before by the consumer. By saving when we call on_demand, the consumers request to set on_demand is not lost when on_hand is called.

3. **explaining** this change: I got into this because the initial version was not working correctly. With  stock_item.send count_on_hand= the stock item data was not saved correctly on mass assignments. I believe it's acceptable to say that these stock management methods on_hand and on_demand require a lock on update and thus a save (this is what spree does on stock_item.count_on_hand). one advantage of this approach is that it makes both on_hand and on_demand behave similarly. So, because it's stock management, I actually think it's a nice exception to the rule of letting the consumers control persistence.

4. **improvements** to this: I believe that if we keep these saves in on_hand and on_demand we make it better by removing attribute_accessible, i.e., these methods need to be called explicitly by the consumers.

3. I believe this the best of the **alternatives** we have so far. what other alternatives do we have? any ideas? maybe some kind of after_save hook on variant to persist the stock_item? I am not sure...

Reply:
Good thoughts. Saving the stock items after saving the variant seems reasonable to me. Worth a try?

There is a difference between on_hand and on_demand in the way they have to be synchronised. on_demand is set by a manager and not modified during checkout. It has only two values. Locking is not essential.

The on_hand value is very critical though. And offering an admin interface to set stock levels is ambiguous. What if the manager counts the product on the shelf and finds that there are `5` while the database still shows `6` (someone was hungry last night). Now the manager enters `5` while a customer buys one at the same time. Should we save `5` or `4`? I think the new Spree 2 approach is wise in allowing only stock increase and decrease, no direct setting. It would end up with `4`.

As long as we keep this consistent (`new_level - stock_item.count_on_hand` uses the two values from the same point in time) it doesn't matter when we save it to the database.

Reply:
>> Saving the stock items after saving the variant seems reasonable to me. 

Rethinking about it. I think the solution in this PR is better. It's more explicit than having callbacks.

Another approach could be to remove on_hand from mass assigments and keep on_demand as mass assignable. That would make the API a bit weird as users would have to save after on_demand but not after calling on_hand.
I think in this case, this would be the normal usage:
variant.on_demand = false
variant.on_hand = 4
variant.save
But I am not sure how to make this sequence work correctly:
variant.on_demand = false
variant.save
variant.on_hand = 4

Do you guys think that would be better than the current PR?

Reply:
I just read both of your comments and let's go with this PR. After the upgrade there will be for sure a cleanup phase where we'll realize about the long term implications of all these small trade-offs #iterate

Thanks for making thoughtfull and objective discussions. We rock :muscle:!

Reply:
ok, cool! This moves us to 90 broken tests in the 2-0-stable build.